### PR TITLE
Organize moved photos into capture-date folders

### PR DESCRIPTION
**Pull request by jss367**
Adds a Quick Move folder-organization choice with common and custom strftime formats, plus previews for the resulting date folders. The backend plans and safely moves tracked photos and companion files across multiple local or mounted-drive destinations, using capture time with mtime fallback and `unsorted` when no usable date exists. Existing same-name files are preserved, remote SSH moves remain single-folder only, and API validation prevents unsafe or conflicting template inputs. Tests cover the UI wiring, multi-folder preflight and job behavior, sidecar moves, catalog updates, and undated photos; `pytest -q vireo/tests/test_move.py vireo/tests/test_move_api.py vireo/tests/test_app.py -q` passes.

## Summary by CodeRabbit

- **New Features**
  - Quick Move can now organize moved photos into capture-date subfolders using a selectable date template (preset or custom).
  - The preview shows resulting folders, destination counts, and an “Organize by Capture Date” confirmation; photos without usable capture dates are placed in **unsorted**.
  - Developed renders are relocated along with moved photos to the appropriate date destination.
- **Bug Fixes**
  - Move cleanup is more resilient: relocation and source-delete issues are surfaced per photo without aborting the batch.
  - Improved safety/validation prevents moving into invalid or non-directory targets during date organization.
- **Limitations**
  - Capture-date organization remains unavailable for remote SSH destinations.

### CODE REVIEW

**Comment by jss367** [comment]
@codex

File: vireo/app.py [hunk 1]
```diff
@@ -19658,16 +19681,25 @@ def progress_cb(current, total, filename, phase="Moving folder"):
...
-                remote=remote,
-                destination_name=destination_name,
-            )
+            if folder_template:
+                result = move_folder_by_date(
+                    db=thread_db,
+                    folder_id=folder_id,
+                    destination=destination,
+                    folder_template=folder_template,
+                    progress_cb=progress_cb,
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/app.py]: Addressed in e98d7948 by threading the configured developed-output directory through date-organized moves and rebasing each moved render.

File: vireo/move.py [hunk 2]
```diff
@@ -1673,12 +1827,47 @@ def move_photos(db, photo_ids, destination, progress_cb=None):
...
+            # Rebase this photo's developed-output file for the new folder
+            # key BEFORE removing originals. `developed_folder_key` hashes the
+            # folder path, so the folder_id update above just invalidated the
+            # lookup — without this rebase, `_iter_developed_outputs` would
+            # look under the new hash, miss the render, and fall back to the
+            # RAW. Doing it before cleanup means a subsequent os.remove
+            # failure (read-only source dir, locked file on Windows) still
+            # leaves the catalog and developed render pointing at the same
+            # (new) folder key.
+            if developed_dir:
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/move.py:2283]: Addressed in 3faec9c7 by relocating the default source-folder developed layout during date-organized moves.

File: vireo/export.py [hunk 3]
```diff
@@ -592,6 +592,157 @@ def relocate_developed_dir(developed_dir, old_folder_path, new_folder_path):
...
+    for name in names:
+        entry_stem = os.path.splitext(name)[0]
+        # Stem match is case-sensitive to match the read side in
+        # `_DevelopedDirIndex`, which uses case-sensitive stems so two
+        # photos differing only in case on a case-sensitive filesystem
+        # don't collide onto each other's developed render.
+        if entry_stem != stem:
+            continue
+        src_file = os.path.join(old_subdir, name)
+        if not os.path.exists(src_file):
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/export.py:648]: Addressed in 23b886c9 and 464beeee by preserving shared-stem renders across different date destinations, with coverage for both developed-output layouts.

File: vireo/export.py [hunk 4]
```diff
@@ -592,6 +592,207 @@ def relocate_developed_dir(developed_dir, old_folder_path, new_folder_path):
...
+            # A prior per-photo call may already have moved this shared-stem
+            # render (e.g. a RAW+JPEG pair). If the rows fan out to different
+            # date folders, the later destination needs its own copy; the
+            # render is only already in place when both rows share a target.
+            if listing_cache is not None:
+                prior_destination = listing_cache.get(relocation_key)
+            if not prior_destination or not os.path.isfile(prior_destination):
+                continue
+        dst_file = os.path.join(new_subdir, name)
+        if os.path.exists(dst_file):
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/export.py:658]: Addressed in 42da9130 by rejecting ambiguous distinct-source developed-render collisions before catalog rows are moved.

File: vireo/app.py [hunk 5]
```diff
@@ -19836,6 +19884,34 @@ def api_move_folder_preflight():
...
+                    _get_db(), folder_id, destination, folder_template,
+                )
+            except ValueError as exc:
+                return json_error(str(exc))
+            destinations = [
+                {
+                    "path": item["destination"],
+                    "relative_path": item["relative_path"],
+                    "photo_count": item["photo_count"],
+                    "exists": os.path.isdir(item["destination"]),
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/app.py:20511]: Addressed in 4fc7ba4e and de8cb3e8 by validating planned date destinations and returning a structured API error when a path is occupied by a file.

File: vireo/export.py [hunk 6]
```diff
@@ -592,6 +592,207 @@ def relocate_developed_dir(developed_dir, old_folder_path, new_folder_path):
...
+                "Skipping developed relocate %s -> %s: destination exists",
+                src_file, dst_file,
+            )
+            continue
+        try:
+            os.makedirs(new_subdir, exist_ok=True)
+            if prior_destination:
+                shutil.copy2(prior_destination, dst_file)
+            elif preserve_source:
+                shutil.copy2(src_file, dst_file)
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/export.py:690]: Addressed in ca5156e0, with follow-up hardening in 385535c7, by cleaning partial relocation copies while preserving complete copies whose source unlink failed.

File: vireo/move.py [hunk 7]
```diff
@@ -1558,21 +1559,274 @@ def _tracked_destination_ancestor(db, folder_id, dest_path):
...
+            descendant_params = ()
+        else:
+            descendant_predicate = f"substr({path_expr}, 1, ?) = ?"
+            descendant_params = (len(prefix), prefix)
+    photos = db.conn.execute(
+        f"""SELECT p.*, f.path AS folder_path
+           FROM photos p
+           JOIN folders f ON f.id = p.folder_id
+           WHERE f.id = ?
+              OR {descendant_predicate}
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/move.py:1660]: Addressed in 8f10881d by using alias-aware descendant matching for date-move planning on POSIX.

File: vireo/move.py [hunk 8]
```diff
@@ -1558,21 +1559,295 @@ def _tracked_destination_ancestor(db, folder_id, dest_path):
...
+        # ``2026/07``; if ``destination/2026`` is already a regular file (or
+        # a dangling symlink), the final candidate does not lexist yet and
+        # only the leaf check misses it, but ``os.makedirs`` in the worker
+        # would then raise ``NotADirectoryError``/``FileExistsError`` and
+        # abort the background job with an opaque exception instead of
+        # returning a structured preflight error.
+        parts = relative.split("/")
+        for depth in range(1, len(parts) + 1):
+            ancestor = os.path.join(destination, *parts[:depth])
+            if os.path.lexists(ancestor) and not os.path.isdir(ancestor):
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/move.py:1760]: Addressed in 2870960b by validating the destination root itself as a directory before planning child paths.

File: vireo/move.py [hunk 9]
```diff
@@ -1610,12 +1933,34 @@ def move_photos(db, photo_ids, destination, progress_cb=None):
...
+            # not by the original extension. Same-stem photos from one source
+            # folder intentionally share a render (RAW+JPEG), but two source
+            # folders can hold distinct renders with the same filename. Do not
+            # merge the latter into one destination and silently make one row
+            # display/export the other's edit.
+            existing_origin = destination_stem_origins.get(
+                stem, no_destination_stem,
+            )
+            if existing_origin is not no_destination_stem \
+                    and existing_origin != src_dir:
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/move.py]: Addressed in f7c8f16a by clearing stored move provenance when a source folder is deleted, so reused paths cannot inherit it.

File: vireo/db.py [hunk 10]
```diff
@@ -4827,6 +4856,19 @@ def delete_folder(self, folder_id):
...
+            # unrelated folder appearing at one of these deleted paths bypass
+            # the same-stem developed-render collision guard in
+            # ``move_photos``. Run after the folder DELETEs (nothing left in
+            # this transaction can re-populate it) and inside the same outer
+            # transaction so a rollback restores both together.
+            for chunk in _chunks(deleted_folder_paths):
+                placeholders = ",".join("?" for _ in chunk)
+                self.conn.execute(
+                    f"UPDATE photos SET last_move_source_folder_path = NULL "
+                    f"WHERE last_move_source_folder_path IN ({placeholders})",
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/db.py:5063]: Addressed in fb719390 by cascading provenance updates through whole-folder path rebases.

File: vireo/move.py [hunk 11]
```diff
@@ -1668,17 +2013,92 @@ def move_photos(db, photo_ids, destination, progress_cb=None):
...
+            except ImportError:
+                from export import (
+                    relocate_default_developed_file,
+                    relocate_developed_file,
+                )
+            source_stem_key = (photo["folder_id"], stem)
+            source_stem_counts[source_stem_key] = max(
+                0, source_stem_counts.get(source_stem_key, 1) - 1,
+            )
+            preserve_source_render = source_stem_counts[source_stem_key] > 0
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/move.py:2248]: Addressed in 1de38827 by expiring provenance as soon as the last same-stem source row moves out.

File: vireo/move.py [hunk 12]
```diff
@@ -1558,21 +1559,309 @@ def _tracked_destination_ancestor(db, folder_id, dest_path):
...
+    # Reject a destination root that is itself a regular file or dangling
+    # symlink. The ancestor walk below only checks paths *inside*
+    # ``destination`` (``depth`` starts at 1), so without this hoisted
+    # check a template like ``%Y/%m`` against ``destination='/archive'``
+    # where ``/archive`` is a plain file would pass preflight — the
+    # rendered leaf ``/archive/2026/07`` does not lexist yet — and
+    # ``os.makedirs`` in the worker would then raise
+    # ``NotADirectoryError``/``FileExistsError`` instead of returning
+    # the structured preflight error this guard is meant to provide.
+    if os.path.lexists(destination) and not os.path.isdir(destination):
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/move.py:1704]: Addressed in 30ba53b9 by validating existing parents above a not-yet-created destination root.

File: vireo/db.py [hunk 13]
```diff
@@ -4827,6 +4879,19 @@ def delete_folder(self, folder_id):
...
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/db.py:5063]: Addressed in 30ba53b9 by clearing provenance in the missing-folder merge deletion path.

File: vireo/db.py [hunk 14]
```diff
@@ -1330,6 +1331,23 @@ def _create_tables(self):
...
+        # ``delete_folder``, so a stale id could compare equal to an unrelated
+        # new folder and bypass the collision guard.
+        try:
+            self.conn.execute(
+                "SELECT last_move_source_folder_path FROM photos LIMIT 0"
+            )
+        except sqlite3.OperationalError:
+            self.conn.execute(
+                "ALTER TABLE photos "
+                "ADD COLUMN last_move_source_folder_path TEXT"
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/db.py:1363]: Addressed in 7a0ff6b2 by rebasing provenance when missing folders are relocated.

File: vireo/move.py [hunk 15]
```diff
@@ -1610,12 +1955,34 @@ def move_photos(db, photo_ids, destination, progress_cb=None):
...
+            # not by the original extension. Same-stem photos from one source
+            # folder intentionally share a render (RAW+JPEG), but two source
+            # folders can hold distinct renders with the same filename. Do not
+            # merge the latter into one destination and silently make one row
+            # display/export the other's edit.
+            existing_origin = destination_stem_origins.get(
+                stem, no_destination_stem,
+            )
+            if existing_origin is not no_destination_stem \
+                    and existing_origin != src_dir:
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/move.py]: Addressed in 7a0ff6b2 by folding destination stem keys according to destination filesystem case sensitivity.

File: vireo/export.py [hunk 16]
```diff
@@ -592,6 +592,237 @@ def relocate_developed_dir(developed_dir, old_folder_path, new_folder_path):
...
+            # A prior per-photo call may already have moved this shared-stem
+            # render (e.g. a RAW+JPEG pair). If the rows fan out to different
+            # date folders, the later destination needs its own copy; the
+            # render is only already in place when both rows share a target.
+            if listing_cache is not None:
+                prior_destination = listing_cache.get(relocation_key)
+            if not prior_destination or not os.path.isfile(prior_destination):
+                continue
+        dst_file = os.path.join(new_subdir, name)
+        if os.path.exists(dst_file):
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/export.py:658]: Addressed in 39d510ee by detecting stale destination developed renders before repointing the catalog row.

File: vireo/db.py [hunk 17]
```diff
@@ -3687,6 +3728,22 @@ def _merge_into_existing(self, source_folder_id, target_folder_id, new_path, *,
...
+        # same-stem developed-render collision guard in ``move_photos``
+        # compares the stored origin to a candidate move's ``src_dir``
+        # by string equality and would let two unrelated destination rows
+        # share developed-output lookup by folder+stem. Runs inside the
+        # merge's transaction so a rollback restores both together.
+        if old_path:
+            self.conn.execute(
+                "UPDATE photos SET last_move_source_folder_path = NULL "
+                "WHERE last_move_source_folder_path = ?",
+                (old_path,),
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/db.py:3864]: Addressed in 39d510ee by cascading provenance cleanup for child folders merged beneath a relocated missing root.

File: vireo/move.py [hunk 18]
```diff
@@ -1558,21 +1559,390 @@ def _tracked_destination_ancestor(db, folder_id, dest_path):
...
+        try:
+            names = os.listdir(subdir)
+        except OSError:
+            continue
+        # Case-sensitive stem comparison mirrors ``_DevelopedDirIndex``:
+        # two photos differing only by case must not collapse onto each
+        # other's render on case-sensitive filesystems.
+        for name in names:
+            if os.path.splitext(name)[0] != stem:
+                continue
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/move.py:1895]: Addressed in 0da868f9: stale-render stem matching now case-folds when the destination filesystem is case-insensitive, with regression coverage.

File: vireo/move.py [hunk 19]
```diff
@@ -1668,17 +2118,126 @@ def move_photos(db, photo_ids, destination, progress_cb=None):
...
+                        (src_dir,),
+                    )
+                    if os.path.splitext(row["filename"])[0] == stem
+                ]
+                if stale_ids:
+                    placeholders = ",".join("?" for _ in stale_ids)
+                    db.conn.execute(
+                        f"UPDATE photos SET "
+                        f"last_move_source_folder_path = NULL "
+                        f"WHERE id IN ({placeholders})",
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/move.py:2278]: Addressed in 0da868f9: delete_photos now expires provenance when deletion drains the final source sibling, in the same transaction, with end-to-end regression coverage.

**Comment by jss367** [comment]
🤖 Routine invoked for an `address-review` task on this PR, but the referenced review had an empty body — it appears to be a wrapper around reply comments acknowledging previously-addressed Codex findings. No new feedback to act on, so no changes were pushed. If you'd like specific unresolved threads (there are still several from the earlier Codex reviews marked `is_resolved: false`) tackled, let me know which and I'll pick them up.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: vireo/move.py [hunk 20]
```diff
@@ -1668,17 +2122,126 @@ def move_photos(db, photo_ids, destination, progress_cb=None):
...
+                        stale_ids,
+                    )
+                    db.conn.commit()
+                destination_stem_origins[_stem_key(stem)] = None
+            if developed_dir:
+                relocate_developed_file(
+                    developed_dir, src_dir, destination, stem,
+                    developed_listing_cache, preserve_source_render,
+                )
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/move.py:2287]: Addressed in c7c2fb42: the selected-photo job now loads the effective darktable output directory and passes it into move_photos, with API regression coverage.

File: vireo/move.py [hunk 21]
```diff
@@ -1558,21 +1559,393 @@ def _tracked_destination_ancestor(db, folder_id, dest_path):
...
+    # volumes (Windows, default macOS APFS) a keyed-by-raw-case dict would
+    # treat ``IMG`` and ``img`` as distinct entries and let two unrelated
+    # source folders' ``IMG.CR3`` + ``img.NEF`` both land in one date folder,
+    # where their ``*.jpg`` renders would collide but this guard wouldn't
+    # detect it. Probe once and fold the stem before every lookup/write into
+    # ``destination_stem_origins``.
+    dest_case_insensitive = _is_case_insensitive_path(destination)
+
+    def _stem_key(raw_stem):
+        return raw_stem.casefold() if dest_case_insensitive else raw_stem
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/move.py]: Addressed in c7c2fb42: render stem keys now fold when either the photo destination or configured developed-output volume is case-insensitive, with mixed-volume regression coverage.

File: vireo/move.py [hunk 22]
```diff
@@ -1558,21 +1559,395 @@ def _tracked_destination_ancestor(db, folder_id, dest_path):
...
+                progress_cb(_start + current, total, filename,
+                            "Organizing by capture date")
+
+        result = move_photos(
+            db,
+            group["photo_ids"],
+            group["destination"],
+            progress_cb=group_progress,
+            developed_dir=developed_dir,
+            developed_listing_cache=developed_listing_cache,
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/move.py:1827]: Addressed in adc3733a: move_photos now preserves a shared XMP while same-stem source rows remain, reuses a verified XMP copy within one destination batch, and lets each later date group copy it before final cleanup. Regression tests cover same- and cross-destination siblings.

File: vireo/move.py [hunk 23]
```diff
@@ -1558,21 +1559,395 @@ def _tracked_destination_ancestor(db, folder_id, dest_path):
...
+                        f"date destination already exists and is not a "
+                        f"directory: {ancestor}"
+                    )
+                break
+            previous = ancestor
+            ancestor = os.path.dirname(ancestor)
+    plans = []
+    for relative in sorted(groups):
+        candidate = os.path.join(destination, *relative.split("/"))
+        resolved = os.path.realpath(candidate)
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/move.py:1734]: Addressed in adc3733a: rendered relative paths are normalized with POSIX semantics before grouping and joining; dot-only output is rejected. Regression tests cover both normalization and rejection.

File: vireo/db.py [hunk 24]
```diff
@@ -3783,6 +3824,22 @@ def _merge_into_existing(self, source_folder_id, target_folder_id, new_path, *,
...
+            )
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/db.py:3865]: Addressed in 1c0da35f: every child folder cascaded by a missing-root merge now rebases last_move_source_folder_path from its old path to its live new path in the same transaction. The regression covers the sibling move and reuse of the freed old child path.

File: vireo/move.py [hunk 25]
```diff
@@ -1558,21 +1559,404 @@ def _tracked_destination_ancestor(db, folder_id, dest_path):
...
+    # non-ASCII stems (e.g. `C:\Älbum`).
+    if sys.platform == "win32":
+        db.conn.create_function(
+            "LOWER_UNICODE", 1,
+            lambda s: s.lower() if s is not None else None,
+        )
+        root = folder["path"].replace("\\", "/").rstrip("/").lower()
+        prefix = root + "/"
+        path_expr = "LOWER_UNICODE(REPLACE(f.path, '\\', '/'))"
+        descendant_predicate = f"substr({path_expr}, 1, ?) = ?"
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/move.py]: Addressed in 6b031f14: the Windows folded predicate now includes exact path aliases as well as descendants. The regression now covers the selected row, a differently-cased child, and a separate exact alias row.

File: vireo/move.py [hunk 26]
```diff
@@ -1610,12 +2034,59 @@ def move_photos(db, photo_ids, destination, progress_cb=None):
...
+            # not by the original extension. Same-stem photos from one source
+            # folder intentionally share a render (RAW+JPEG), but two source
+            # folders can hold distinct renders with the same filename. Do not
+            # merge the latter into one destination and silently make one row
+            # display/export the other's edit.
+            existing_origin = destination_stem_origins.get(
+                _stem_key(stem), no_destination_stem,
+            )
+            if existing_origin is not no_destination_stem \
+                    and existing_origin != src_dir:
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/move.py]: Addressed in 88eb56a8: destination provenance now records the exact source stem alongside the folded lookup key, so same-source sharing is allowed only for exact-stem siblings. A regression covers IMG and img in one case-sensitive source moving to a folding destination.

File: vireo/move.py [hunk 27]
```diff
@@ -1610,12 +2041,61 @@ def move_photos(db, photo_ids, destination, progress_cb=None):
...
+            # untracked ``<destination>/developed/<stem>.*`` or
+            # ``<developed_dir>/<developed_folder_key(destination)>/<stem>.*``
+            # would be silently served as this photo's developed output
+            # after the row is repointed. Treat it as a move collision
+            # before touching the row.
+            if existing_origin is no_destination_stem and \
+                    _has_untracked_destination_developed(
+                        destination, stem, developed_dir,
+                        case_insensitive=render_case_insensitive,
+                    ):
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/move.py:2130]: Addressed in 4b6acb34: the stale-render probe now receives the current source file and exempts that exact inode, so a tracked source folder named developed can move into its parent. The regression exercises the full catalog and filesystem move.

File: vireo/move.py [hunk 28]
```diff
@@ -1558,21 +1559,417 @@ def _tracked_destination_ancestor(db, folder_id, dest_path):
...
+    if not destination or not stem:
+        return False
+    subdirs = [os.path.join(destination, "developed")]
+    if developed_dir:
+        subdirs.append(
+            os.path.join(developed_dir, developed_folder_key(destination))
+        )
+    for subdir in subdirs:
+        if not subdir or not os.path.isdir(subdir):
+            continue
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/move.py:1884]: Fixed in 59164164. `move_photos` now checks both default and configured destination developed-output paths with `lexists` before any catalog row is inserted or repointed, returning structured per-photo errors when a regular file or dangling symlink blocks relocation. Added a regression test proving the original and catalog row remain untouched.

**Comment by jss367** [comment]
Babysitting has reached the configured cap of 8 autonomous feedback rounds. One new unresolved P2 remains: `last_move_source_folder_path` can become reusable provenance if a removable source disappears and later the same mount path contains different media, potentially allowing a same-stem photo to inherit the prior destination render. I have left that thread unresolved and have not started a ninth code repair round. Separately, CI is rerunning an Ubuntu-only timing failure in `test_api_new_images_response_trims_sample_but_caches_full_list`; the focused test passes locally and every other platform/check is green. Human direction is needed on whether to authorize another repair round for the remaining provenance edge case.

**Comment by jss367** [comment]
🤖 Acknowledged — standing down. This comment fired the address-comment routine, but it reads as a status update explicitly asking for human direction ("Human direction is needed on whether to authorize another repair round"), not as a fix request. Starting a ninth autonomous repair round would defy the stated 8-round cap, so I have not touched code.

The remaining P2 (`last_move_source_folder_path` as reusable provenance when a removable source disappears and a different mount later reuses the path) stays unresolved and unclaimed. If you want me to take a swing at it, drop a follow-up comment authorizing another round and I'll pick it up.

---
_Generated by [Claude Code](https://claude.ai/code)_
